File: Golang/677. Map Sum Pairs.go
type MapSum struct {
    dict map[string]int
}


/** Initialize your data structure here. */
func Constructor() MapSum {
    return MapSum{make(map[string]int)}
}


func (this *MapSum) Insert(key string, val int)  {
    this.dict[key] = val
}


func (this *MapSum) Sum(prefix string) int {
    ans := 0
    
    // length := len(prefix)
    // for key := range this.dict {
    //     if len(key) >= length && key[: length] == prefix {
    //         ans += this.dict[key]
    //     }
    // }
    
    for k, v := range this.dict {
        if strings.HasPrefix(k, prefix) {
            ans += v
        }
    }
    return ans
}


/**
 * Your MapSum object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(key,val);
 * param_2 := obj.Sum(prefix);
 */

File: Python3/677. Map Sum Pairs.py
class MapSum:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.Map = collections.defaultdict(int)
        

    def insert(self, key, val):
        """
        :type key: str
        :type val: int
        :rtype: void
        """
        self.Map[key] = val
        

    def sum(self, prefix):
        """
        :type prefix: str
        :rtype: int
        """
        return sum([v for k, v in self.Map.items() if k.startswith(prefix)])
                
        


# Your MapSum object will be instantiated and called as such:
# obj = MapSum()
# obj.insert(key,val)
# param_2 = obj.sum(prefix)
